src/gennylib/include/gennylib/quiesce.hpp

```diff
@@ -0,0 +1,138 @@
+// Copyright 2021-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef HEADER_5936639A_7D22_4629_8FE1_4A08443DDB0F_INCLUDED
+#define HEADER_5936639A_7D22_4629_8FE1_4A08443DDB0F_INCLUDED
+
+#include <string_view>
+
+#include <mongocxx/uri.hpp>
+#include <mongocxx/client.hpp>
+#include <mongocxx/database.hpp>
+#include <bsoncxx/types.hpp>
+#include <bsoncxx/array/view.hpp>
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/builder/basic/kvp.hpp>
+
+#include <boost/log/trivial.hpp>
+
+#include <gennylib/topology.hpp>
+
+namespace genny {
+
+const int DROPPED_COLLECTION_RETRIES = 1000;
+
+bool waitOplog(Topology& topology) {
+    class WaitOplogVisitor : public TopologyVisitor {
+    public:
+        void visitReplSetDescriptionPre(const ReplSetDescription& desc) {
+            using bsoncxx::builder::basic::kvp;
+            using bsoncxx::builder::basic::make_document;
+            mongocxx::client client(mongocxx::uri(desc.primaryUri));
+            auto admin = client["admin"];
+
+            // Assert that all nodes are in reasonable states.
+            auto replSetRes = admin.run_command(make_document(kvp("replSetGetStatus", 1)));
+            auto members = replSetRes.view()["members"];
+            if (members && members.type() == bsoncxx::type::k_array) {
+                bsoncxx::array::view members_view = members.get_array();
+                for (auto member : members_view) {
+                    std::string state(member["stateStr"].get_utf8().value);
+                    if (state != "PRIMARY" && state != "SECONDARY" && state != "ARBITER") {
+                        std::string name(member["name"].get_utf8().value);
+                        BOOST_LOG_TRIVIAL(error) << "Cannot wait oplog, replset member "
+                            << name << " is " << state;
+                        success_acc = false;
+                        return;
+                    }
+                }
+            }
+
+            // Do flush
+            auto collection = admin["wait_oplog"];
+            mongocxx::write_concern wc;
+            wc.nodes(desc.nodes.size());
+            wc.journal(true);
+            mongocxx::options::insert opts;
+            opts.write_concern(wc);
+            auto insertRes = collection.insert_one(make_document(kvp("x", "flush")), opts);
```

Comment:
I suspect you're just copying the js impl here and that there's a reason for the `{x:flush}` doc and whatnot, but this seems like voodoo. Can you either add a comment about what this is doing from a conceptual level or add a permalink to the js impl?

Reply:
Yeah will do! The js code isn't explained either but my understanding is to force the oplog to flush by doing a write with `w: all` and `j: true` so when that write is finished you know everything before it is replicated.

There's also [awaitReplication()](https://github.com/mongodb/mongo/blob/r4.1.4/src/mongo/shell/replsettest.js#L1354) in the shell which I think does something similar but its implementation is _way_ more complicated.